File: src/main/kotlin/com/wynnlab/spells/archer/ArrowRain.kt
package com.wynnlab.spells.archer

import com.wynnlab.spells.SpellL
import com.wynnlab.spells.SpellData
import com.wynnlab.util.RAD2DEG
import com.wynnlab.util.plus
import org.bukkit.Particle
import org.bukkit.Sound
import org.bukkit.entity.Player
import org.bukkit.util.Vector
import kotlin.math.cos
import kotlin.math.sin

class ArrowRain(player: Player) : SpellL(player, 11, SpellData.ARROW_SHIELD /*TODO*/) {
    override fun tick() {
        when {
            t < 10 -> {
                for (i in 0..2) {
                    val l = player.location.plus(
                        sin((10.0 * t + 60.0 * (i - 1.0)) * RAD2DEG) * (1.5 - .15 * t),
                        t.toDouble(),
                        cos((10.0 * t + 60.0 * (i - 1.0)) * RAD2DEG) * (1.5 - .15 * t)
                    )
                    player.spawnParticle(if (clone) Particle.FIREWORKS_SPARK else Particle.CRIT, l, 5, .0, .0, .0, if (clone) .1 else .3)
                    player.spawnParticle(Particle.SQUID_INK, l, 3, .0, .0, .0, .0)
                    player.playSound(l, Sound.ENTITY_ARROW_SHOOT, 1f, 1f + t / 20f)
                }
            }
            else -> {
                val l = player.location.plus(.0, 10.0, .0)
                player.playSound(l, Sound.ENTITY_ARROW_SHOOT, 1f, .8f)
                player.playSound(l, if (clone) Sound.ITEM_ARMOR_EQUIP_GOLD else Sound.ENTITY_GENERIC_EXPLODE, .8f, if (clone) 1f else 1.8f)
                if (clone) player.playSound(l, Sound.ITEM_TOTEM_USE, .8f, .8f)
                for (i in 0..360 step 30) {
                    for (j in 0..2) {
                        val arrow = player.world.spawnArrow(l, Vector(sin(i * RAD2DEG), .5 * (j - 1.0) - 1.0, cos(i * RAD2DEG)), 3f, 1f)
                        arrow.shooter = player
                        arrow.addScoreboardTag("rain_arrow")
                    }
                }
            }
        }
    }
}

File: src/main/resources/classes/archer/arrow_rain.py
from org.bukkit import Material, Particle, Sound
from org.bukkit.entity import EntityType, Player
from org.bukkit.inventory import ItemStack
from org.bukkit.util import Vector

from com.wynnlab.spells import PySpell

from java.lang import Math

class Spell(PySpell):
    def tick(self):
        if self.t < 10:
            for i in range(3):
                l = self.player.getLocation().clone().add(Math.sin((10 * self.t + 60 * (i - 1)) * DEG2RAD) * (1.5 - .15 * self.t), self.t, Math.cos((10 * self.t + 60 * (i - 1)) * DEG2RAD) * (1.5 - .15 * self.t))
                self.particle(l, Particle.FIREWORKS_SPARK if self.clone else Particle.CRIT, 5, 0, 0, 0, .1 if self.clone else .3)
                self.particle(l, Particle.SQUID_INK, 3, 0, 0, 0, 0)
                self.sound(l, Sound.ENTITY_ARROW_SHOOT, 1, 1 + self.t / 20)
            return

        l = self.player.getLocation().clone().add(0, 10, 0)
        self.sound(l, Sound.ENTITY_ARROW_SHOOT, 1, .8)
        self.sound(l, Sound.ITEM_ARMOR_EQUIP_GOLD if self.clone else Sound.ENTITY_GENERIC_EXPLODE, .8, 1 if self.clone else 1.8)
        if self.clone:
            self.sound(l, Sound.ITEM_TOTEM_USE, .8, .8)

        for i in range(0, 360, 30):
            for j in range(3):
                arrow = self.player.getWorld().spawnArrow(l, Vector(Math.sin(i * DEG2RAD), .5 * (j - 1) - 1, Math.cos(i * DEG2RAD)), 3, 1)
                if self.clone:
                    snowball = self.player.getWorld().spawnEntity(arrow.getLocation(), EntityType.SNOWBALL)
                    snowball.setVelocity(arrow.getVelocity())
                    snowball.setItem(ItemStack(Material.FLINT))
                    arrow.remove()
                    arrow = snowball

                arrow.setShooter(self.player)
                arrow.addScoreboardTag('rain_arrow')

def delete_arrow(event):
    hit = event.getHitEntity()
    if not hit is None and (isinstance(hit, Mob) or isinstance(hit, Player) and hit.getScoreboardTags().contains('pvp') and event.getEntity().getShooter().getScoreboardTags().contains('pvp')):
        PySpell.damage(event.getEntity().getShooter(), hit, False, 2, .7, 0, 0, 0, 0, .3)
        PySpell.knockbackFromPlayer(hit, event.getEntity().getShooter(), .5)

    event.getEntity().remove()

PySpell.registerProjectileHit('rain_arrow', delete_arrow)